UI: add LinkButton (#78944)
WordPress/gutenberg: packages/ui/src/button/stories/usage-guidelines.mdx

Reviewer: Love this doc, thanks for adding it!

Reviewer 2: > Love this doc, thanks for adding it!

I do too! It makes me wonder how we can get this in the Storybook manifest and accessible to [the design system MCP server](https://github.com/WordPress/gutenberg/tree/trunk/packages/design-system-mcp) so that AI agents have access to it. I don't know that it's as easy as just adding a `tags` somewhere. I think this falls under [the docs manifest](https://storybook.js.org/docs/ai/manifests#docs-mdx-manifest) which I guess [we don't emit currently](https://wordpress.github.io/gutenberg/manifests/docs.json)?

Author: Thanks! Overall I'm looking forward to having more design guidance or "how to use the system and why" -docs in addition to more mechanical "what the component does". There are a few already, but literally every component could be paired with these.

import { Canvas, Markdown, Meta } from '@storybook/addon-docs/blocks';
import * as UsageGuidelinesStories from './usage-guidelines.story';

<Meta of={ UsageGuidelinesStories } />

> # Button, IconButton, Link, and LinkButton

Reviewer: One nuance that comes up in modern web apps are actions that change the URL but don't push an item onto the history stack. A common example would be a button that opens some kind of dialog or panel, and that new view has a permalink that is `replaceStack()`ed (instead of `push`) to the history stack. We actually need to use button semantics in this case. "Navigate" is a somewhat ambiguous term in an SPA-context to begin with, so the most accurate criteria for choosing between button/link is going to be "does it **push** a new item onto the history stack".

It's perhaps a bit of an advanced concept to use as our basic guideline, but we might want to capture this nuance _somewhere_ in this doc as well.

Author: Yeah; I also don't also want to make document too technical sounding since one audience should be designers who necessarily don't know what we mean by technical descriptions like you're suggesting.

Agree that "Navigation" is vague, or the whole topic is vague;
- SPAs can usually navigate with onClick or href just fine
- Buttons opening modals, filter buttons, etc can set state in URL, e.g. `#modal-open`, `?filters=` which can intentionally be part of history stack.

Perhaps there's a way to think and frame "navigate" entirely as non non-technical term, but I'm not sure how to phrase it. Good as-is for now and we can always iterate?

Choose the component based on **what happens when the user activates it**, not only on how it should look.

## Quick reference

<Markdown>{ `
| Goal | Component | Element | Examples |
| --- | --- | --- | --- |

> | Performs an action on the current page. | [Button](?path=/docs/design-system-components-button--docs), [IconButton](?path=/docs/design-system-components-iconbutton--docs) | \`<button>\` | Save, Delete, Submit, toolbar icon actions |

Reviewer: ```suggestion
| Performs an action on the current page | [Button](?path=/docs/design-system-components-button--docs), [IconButton](?path=/docs/design-system-components-iconbutton--docs) | \`<button>\` | Save, Delete, Submit, toolbar icon actions |
```

| Navigate to another page or URL | [Link](?path=/docs/design-system-components-link--docs) | \`<a>\` | "Learn more", docs links, external references |

> | Navigate with button styling | [LinkButton](?path=/docs/design-system-components-linkbutton--docs) | \`<a>\` | Standalone CTAs, card actions |

Reviewer: "card _actions_" reads as a bit ambiguous/contradictory for an example for links. What specifically did we mean here?

Author: This type of card with links seems pretty common:

<img width="431" height="416" alt="image" src="https://github.com/user-attachments/assets/d8b42f9b-99c5-4b3e-bb49-f8d841a383e1" />

Since they're primary CTAs, folks don't want to use Link for visual reasons; a button carries more "CTA feel" and is more noticeable.

Standalone would be more like what could be in EmptyState in the middle of the page:

<img width="357" height="231" alt="image" src="https://github.com/user-attachments/assets/7d756a40-27f5-403e-9f77-6f51244d0354" />

Word are hard, and visual examples help. :-) I want to revisit this guide in a follow-up with clearer do/don't illustrations and "Do/Don't do" sections (similar to [what I did in Badge](https://github.com/WordPress/gutenberg/pull/79585)) , but I think it's better for a follow-up.

` }</Markdown>

## Use `Button` and `IconButton` for actions

Use `Button` or `IconButton` when activation should **not** navigate away as a link would:

-   Submitting or resetting a form (`type="submit"` / `type="reset"`)
-   Opening overlays ([Dialog](?path=/docs/design-system-components-dialog--docs), [Popover](?path=/docs/design-system-components-popover--docs), etc.)
-   Running JavaScript on the current page (`onClick` without a meaningful `href`)
-   Toggle controls (`aria-pressed` on neutral minimal buttons)
-   Showing a **loading** state while an async action completes

Use [IconButton](?path=/docs/design-system-components-iconbutton--docs) for the same kinds of actions when the control is **icon-only** — for example, toolbar buttons. It inherits `Button` props and adds a built-in tooltip via the required `label` prop (used as the accessible name).

> `Button` and `IconButton` use button semantics from Base UI. Do not use them as stand-ins for navigation when a real URL is the primary outcome.

Reviewer: ```suggestion
`Button` and `IconButton` use button semantics. Do not use them as stand-ins for navigation when a real URL is the primary outcome.
```

I generally want to avoid leaking the concept of "Base UI" into our general audience docs, especially when it's not necessary to understand the message we want to convey. Most people don't know what Base UI is, and shouldn't have to know.

<Canvas of={ UsageGuidelinesStories.UseButtonForActions } />

See [Button](?path=/docs/design-system-components-button--docs) and [IconButton](?path=/docs/design-system-components-iconbutton--docs) for API details and examples.

## Use `Link` for navigation

Use `Link` when activation should take the user to another page or URL. This is the default choice for navigation — not just inline references in running text, but any case where the outcome is “go somewhere else.”

`Link` styling communicates navigation more clearly than a button-shaped control. Users can see at a glance that they are following a link, not triggering an action on the current page.

Typical uses:

-   Inline references inside [Text](?path=/docs/design-system-components-text--docs) or paragraphs
-   Standalone “Learn more”, “View documentation”, or “Open site” actions
-   External destinations with `openInNewTab` (adds `target="_blank"` and an accessible new-tab indicator)

> Prefer composing `Text` as the host and `Link` via `render` when you need typography from `Text` on a standalone link.

Reviewer: I believe this needs to be worded more strongly because it's not a matter of preference, it will actually result in incorrect markup if it isn't followed (see [lint rule](https://github.com/WordPress/gutenberg/blob/trunk/packages/eslint-plugin/docs/rules/no-unsafe-render-order.md) for more information).

```suggestion
Compose `Text` as the host and `Link` via `render` when you need typography from `Text` on a standalone link.
```

Do **not** use `LinkButton` for form submit, dialogs, or other non-navigation actions — use `Button` instead.

Do **not** use `Button` with `render={ <a href="…" /> }` and `nativeButton={ false }` for new code when `LinkButton` fits; `LinkButton` is the supported pattern for button-styled links.

<Canvas of={ UsageGuidelinesStories.UseLinkButtonForNavigation } />

> **Note:** Prefer `Link` for navigation when possible. Its underline and link styling communicate where the user is going more clearly than a button-shaped control. Reach for `LinkButton` only when you have considered `Button` and `Link` and still need button prominence.

See [LinkButton](?path=/docs/design-system-components-linkbutton--docs) for API details and examples.

## Common mistakes

-   **`Button` or `IconButton` + `onClick` for navigation** — Prefer `Link` with a real `href` so keyboard, assistive technology, and browser link affordances stay correct.
-   **`LinkButton` when `Link` would work** — Default to `Link` for navigation; button styling can obscure that the control goes elsewhere.

> -   **`LinkButton` for submit or toggle actions** — Use `Button`; only `Button` supports loading and pressed toggle styling.

Reviewer 3: ```suggestion
-   **`LinkButton` for submit or toggle actions** — Use `Button` or `IconButton`, depending on whether the control needs a visible text label or is icon-only; `LinkButton` does not support loading or pressed toggle styling.
```

import { Canvas, Markdown, Meta } from '@storybook/addon-docs/blocks';
import * as UsageGuidelinesStories from './usage-guidelines.story';

<Meta of={ UsageGuidelinesStories } />

# Button, Link, and LinkButton

Choose the component based on **what happens when the user activates it**, not

> only on how it should look.

Reviewer 3: Nit: I don't think we normally break the line in markdown files for formatting reasons?

```suggestion
Choose the component based on **what happens when the user activates it**, not only on how it should look.
```

Here and everywhere else applicable in the PR

Author: good catcha! https://github.com/WordPress/gutenberg/pull/78944/commits/389eca9b85f68578af0922bcead92c166303b6a2

## Quick reference

<Markdown>{`
| Goal | Component | Element | Examples |
| --- | --- | --- | --- |

> | Change something on the current page | [Button](?path=/docs/design-system-components-button--docs) | \`<button>\` | Save, Delete, Open dialog, Toggle |

Reviewer 3: ```suggestion
| Performs an action on the current page. | [Button](?path=/docs/design-system-components-button--docs) | \`<button>\` | Save, Delete, Submit, Open dialog, Toggle |
```

| Navigate to another page or URL | [Link](?path=/docs/design-system-components-link--docs) | \`<a>\` | "Learn more", docs links, external references |

> | Navigate with button styling | [LinkButton](?path=/docs/design-system-components-link-button--docs) | \`<a>\` | Standalone CTAs, card actions |

Reviewer 3: The URL generated by Storybook doesn't have the hyphen

```suggestion
| Navigate with button styling | [LinkButton](?path=/docs/design-system-components-linkbutton--docs) | \`<a>\` | Standalone CTAs, card actions |
```

examples.

## Use `LinkButton` for navigation with button styling

Use `LinkButton` when the destination is a **URL** and the control should
**look like a `Button`** — for example, standalone calls to action, card
footers, or notice actions where button weight fits the layout.

`LinkButton` shares `Button` visual props (`variant`, `tone`, `size`)
but renders an `<a>` like `Link`. It does **not** support `disabled`, `loading`,
`focusableWhenDisabled`, `nativeButton`, or toggle `aria-pressed`.

Do **not** use `LinkButton` for form submit, dialogs, or other non-navigation
actions — use `Button` instead.

Do **not** use `Button` with `render={ <a href="…" /> }` and `nativeButton={
false }` for new code when `LinkButton` fits; `LinkButton` is the supported
pattern for button-styled links.

<Canvas of={ UsageGuidelinesStories.UseLinkButtonForNavigation } />

> **Note:** Prefer `Link` for navigation when possible. Its underline and link
> styling communicate where the user is going more clearly than a button-shaped
> control. Reach for `LinkButton` only when you have considered `Button` and
> `Link` and still need button prominence.

> See [LinkButton](?path=/docs/design-system-components-link-button--docs) for API

Reviewer 3: The URL generated by Storybook doesn't have the hyphen

```suggestion
See [LinkButton](?path=/docs/design-system-components-linkbutton--docs) for API
```

Author: fixed https://github.com/WordPress/gutenberg/pull/78944/commits/ab9eff67d242a001d9b60b62e2e7f1ea24aa145e
